Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,182 @@
+import contextlib
+import enum
+
+from types import TracebackType
+from typing import final, Any, Iterator, Optional, Type
+
+from . import events
+from . import exceptions
+from . import tasks
+
+
+__all__ = (
+    "cancel_scope",
+    "cancel_scope_at",
+    "timeout",
+    "timeout_at",
+    "CancelScope"
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    CANCELLING = "cancelling"
+    CANCELLED = "cancelled"
+    EXITED = "exited"
+
+
+@final
+class CancelScope:
+    __slots__ = ("_deadline", "_loop", "_state", "_task", "_timeout_handler")
```

Comment:
Why slots? To save a few nsec? Or because you're planning a C version? (The latter would make some sense.)

Reply:
I have no strong opinion here. The only thing: `__slots__` can be dropped in a backward-compatible manner, adding a slot can break third-party code.

Reply:
Most public data structures in asyncio don't have slots. E.g. this works -- you can set your own attributes on these if you want to (at your own risk):
```py
f = asyncio.Future()
f.foo = 1
assert f.foo == 1
t = asyncio.Task(...)
t.bar = 2
assert t.bar == 2
```
So I think we should not have slots for timeout either. (Nor do we have slots for TaskGroup.)

Reply:
Ok